### PR TITLE
cluster up: support --host-config-dir

### PR DESCRIPTION
**Pull request by mfojtik**
@deads2k adds `--write-config` that you must set to populate the host-config-dir with configs... if the dir is empty and you run cluster up it will fail.

### CODE REVIEW

**Comment by mfojtik** [comment]
@deads2k still haven't figured out what erases `oc-cluster-up-kube-apiserver` directory after cluster up finishes... other dirs seems to stay but the `oc-cluster-up-kube-apiserver` is removed...

File: pkg/oc/bootstrap/clusterup/kubeapiserver/config.go [hunk 1]
```diff
@@ -10,15 +10,19 @@ import (
...
+const ComponentDirectoryName = "oc-cluster-up-kube-apiserver"
+
 type KubeAPIServerStartConfig struct {
 	// MasterImage is the docker image for openshift start master
 	MasterImage string
 	ImageFormat string
 	DNSPort     int
 	PublicHost  string
+
+	HostDir string
```
Comment (deads2k) [inline pkg/oc/bootstrap/clusterup/kubeapiserver/config.go]: I don't think we need to pass this through.  Skip the entire task if it is specified. If it is not specified, generated into a tempdir and if they want `--write-config`, they can specify it and we'll copy it out.

File: pkg/oc/bootstrap/clusterup/kubelet/config.go [hunk 2]
```diff
@@ -19,6 +24,8 @@ type NodeStartConfig struct {
 	NodeImage string
 
 	Args []string
+
+	HostDir string
```
Comment (deads2k) [inline pkg/oc/bootstrap/clusterup/kubelet/config.go]: same "don't need" comment.

File: pkg/oc/bootstrap/clusterup/kubelet/config.go [hunk 3]
```diff
@@ -28,10 +35,23 @@ func NewNodeStartConfig() *NodeStartConfig {
 
 }
 
+func (opt NodeStartConfig) MakeKubeDNSConfig(dockerClient dockerhelper.Interface, imageRunHelper *run.Runner, out io.Writer) (string, error) {
+	return opt.makeConfig(dockerClient, imageRunHelper, out, ComponentKubeDNSDirectoryName)
```
Comment (deads2k) [inline pkg/oc/bootstrap/clusterup/kubelet/config.go]: Pretty sure that this mutates the kubeconfig server values.

File: pkg/oc/bootstrap/docker/openshift/helper.go [hunk 4]
```diff
@@ -395,7 +395,7 @@ func masterHTTPClient(localConfig string) (*http.Client, error) {
 // copyConfig copies the OpenShift configuration directory from the
 // server directory into a local temporary directory.
 func (h *Helper) copyConfig() (string, error) {
```
Comment (deads2k) [inline pkg/oc/bootstrap/docker/openshift/helper.go]: Do not want this method at all.  Can we leave it unchanged?

File: pkg/oc/bootstrap/docker/run_self_hosted.go [hunk 5]
```diff
@@ -184,6 +194,7 @@ func (c *ClientStartConfig) makeMasterConfig(out io.Writer) (string, error) {
 	container.ImageFormat = c.imageFormat()
 	container.DNSPort = c.DNSPort
 	container.PublicHost = publicHost
+	container.HostDir = c.HostConfigDir
```
Comment (deads2k) [inline pkg/oc/bootstrap/docker/run_self_hosted.go]: extraneous

File: pkg/oc/bootstrap/docker/run_self_hosted.go [hunk 6]
```diff
@@ -201,6 +212,7 @@ func (c *ClientStartConfig) makeNodeConfig(out io.Writer, masterConfigDir string
 	container := kubelet.NewNodeStartConfig()
 	container.ContainerBinds = append(container.ContainerBinds, masterConfigDir+":/var/lib/origin/openshift.local.masterconfig:z")
 	container.NodeImage = c.openshiftImage()
+	container.HostDir = c.HostConfigDir
```
Comment (deads2k) [inline pkg/oc/bootstrap/docker/run_self_hosted.go]: extraneous

File: pkg/oc/bootstrap/docker/run_self_hosted.go [hunk 7]
```diff
@@ -219,7 +231,7 @@ func (c *ClientStartConfig) makeNodeConfig(out io.Writer, masterConfigDir string
 	if err != nil {
 		return "", "", fmt.Errorf("error creating node config: %v", err)
 	}
-	kubeDNSConfigDir, err := container.MakeNodeConfig(c.GetDockerClient(out), imageRunningHelper.New(), out)
+	kubeDNSConfigDir, err := container.MakeKubeDNSConfig(c.GetDockerClient(out), imageRunningHelper.New(), out)
 	if err != nil {
 		return "", "", fmt.Errorf("error creating node config: %v", err)
 	}
```
Comment (deads2k) [inline pkg/oc/bootstrap/docker/run_self_hosted.go]: Below this somewhere, there should be an `if c.WriteConfig{ //copy the content to the correct locations and exit}`

File: pkg/oc/util/dir/util.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+)
+
+// ConfigDir provides directory creation for cluster up components. In case the hostDir is specified, it will use
+// that directory as a base directory for all configs. In case the hostDir already exists, this will return true.
+// In case it is not specified a temporary directory is created.
+func ConfigDir(hostDir, component string) (string, error) {
```
Comment (deads2k) [inline pkg/oc/util/dir/util.go]: This is bounded to cluster-up.  Never auto create anything from this method.  If they mess up, just fail.

**Comment by deads2k** [comment]
merging and I'll mess with it

**Comment by deads2k** [comment]
thanks

**Pull request merged**